pkg/kubelet/prober/prober.go

```diff
@@ -93,29 +93,36 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
 	}
 
 	if probeSpec == nil {
-		klog.InfoS("Probe is nil", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+		klog.V(3).InfoS("Probe is nil", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
 		return results.Success, nil
 	}
 
 	result, output, err := pb.runProbeWithRetries(ctx, probeType, probeSpec, pod, status, container, containerID, maxProbeRetries)
-	if err != nil || (result != probe.Success && result != probe.Warning) {
-		// Probe failed in one way or another.
-		if err != nil {
-			klog.V(1).ErrorS(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
-			pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerUnhealthy, "%s probe errored: %v", probeType, err)
-		} else { // result != probe.Success
-			klog.V(1).InfoS("Probe failed", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "probeResult", result, "output", output)
-			pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerUnhealthy, "%s probe failed: %s", probeType, output)
-		}
+
+	if err != nil {
+		// Handle probe error
+		klog.V(1).ErrorS(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerUnhealthy, "%s probe errored: %v", probeType, err)
 		return results.Failure, err
 	}
-	if result == probe.Warning {
-		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe warning: %s", probeType, output)
-		klog.V(3).InfoS("Probe succeeded with a warning", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
-	} else {
-		klog.V(3).InfoS("Probe succeeded", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+
+	switch result {
+	case probe.Success:
+		return results.Success, nil
+
+	case probe.Warning:
+		klog.V(1).InfoS("Probe warning", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
+		return results.Success, nil
+
+	case probe.Unknown:
```

Comment:
If the probe’s state is Unknown, it will always return an error, so I don’t think this part of the code will ever be executed.
https://github.com/kubernetes/kubernetes/blob/c7e6c419868d131127ceccf61f3c077436f1b3bb/pkg/kubelet/prober/prober.go#L157
https://github.com/kubernetes/kubernetes/blob/c7e6c419868d131127ceccf61f3c077436f1b3bb/pkg/kubelet/prober/prober.go#L174
https://github.com/kubernetes/kubernetes/blob/c7e6c419868d131127ceccf61f3c077436f1b3bb/pkg/kubelet/prober/prober.go#L194

If the probe's state is Unknown, it'll go into this error handling.
https://github.com/kubernetes/kubernetes/blob/c7e6c419868d131127ceccf61f3c077436f1b3bb/pkg/kubelet/prober/prober.go#L102-L107

So, I think it'd be better to go with these changes instead.

```diff
diff --git a/pkg/kubelet/prober/prober.go b/pkg/kubelet/prober/prober.go
index 3b771847bf8..08afa16a164 100644
--- a/pkg/kubelet/prober/prober.go
+++ b/pkg/kubelet/prober/prober.go
@@ -98,7 +98,11 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
        }

        result, output, err := pb.runProbeWithRetries(ctx, probeType, probeSpec, pod, status, container, containerID, maxProbeRetries)
-
+       if err != nil && result == probe.Unknown {
+               klog.V(1).ErrorS(err, "Probe unknown", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+               pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe resulted in unknown state: %v", probeType, err)
+               return results.Failure, err
+       }
        if err != nil {
                // Handle probe error
                klog.V(1).ErrorS(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
@@ -114,10 +118,6 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
                klog.V(1).InfoS("Probe warning", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
                return results.Success, nil

-       case probe.Unknown:
-               klog.V(1).InfoS("Probe unknown", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
-               return results.Failure, nil
-
        default:
                klog.V(1).InfoS("Probe failed", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "probeResult", result, "output", output)
                pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerUnhealthy, "%s probe failed: %s", probeType, output)
```

Reply:
This complicates the code and makes an assumption that `probe.Unknown` will always come with non-nil error. We shouldn't assume this. I'm for keeping @jralmaraz approach.

Reply:
> We shouldn't assume this.

Yeah, it’s fine to keep the current switch case as is. But in the current implementation, an error gets returned every time the probe is unknown, so when `runProbeWithRetries` hits the retry limit, it returns an error. Like I mentioned above, this sends it into the previous error handling section and treats everything as a 'Probe errored'. So, if we want to log and write an event as 'Probe unknown', I think the suggested code addition is necessary.

Reply:
Thanks for the inputs ! I've tried splitting into `Probe unknown with error` and `Probe unknown without error`. Does it help addressing the concern without adding complication to the code ? 

`	if err != nil && result == probe.Unknown {
		klog.V(1).ErrorS(err, "Probe unknown with error", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe resulted in unknown state: %v", probeType, err)
		return results.Failure, err
	}

	if result == probe.Unknown {
		klog.V(1).InfoS("Probe unknown without error", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
		return results.Failure, nil
	}
`

Reply:
I'd prefer the latter to go into `switch` statement below. As for the former, what about including result into the message, so for any result type the error message would be informative enough?

Reply:
> I'd prefer the latter to go into switch statement below.

+1. It's easier to read that way.

```go
	switch result {
	case probe.Success:
		return results.Success, nil

	(...)

	case probe.Unknown:
		klog.V(1).InfoS("Probe unknown without error", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
		return results.Failure, nil

	(...)
	}
```